features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+
+interface TOTPAuthenticator {
+  id: string;
+  type: "totp";
+  activatedAt: Date;
+  displayName: string;
+}
+
+interface OOBAuthenticator {
+  id: string;
+  type: "oob";
+  activatedAt: Date;
+  channel: "sms";
+  phone: string;
+}
+
+interface RecoveryCodeAuthenticator {
+  id: string;
+  type: "recovery_code";
+  createdAt: Date;
+}
+
+interface BearerTokenAuthenticator {
+  id: string;
+  type: "bearer_token";
+  createdAt: Date;
+  expireAt: Date;
+}
+
+interface RegenerateRecoveryCodeResult {
+  recoveryCodes: string[];
+}
+
+async function listAuthenticators(): Promise<Authenticator[]>;
+async function deleteAuthenticator(authenticatorID: string): Promise<void>;
+async function regenerateRecoveryCode(): Promise<RegenerateRecoveryCodeResult>;
+```
+
+# Use Cases
+
+## Registration
+
+```typescript
+try {
+  await skygear.auth.login("user@example.com", "password");
+} catch (e) {
+  if (skygear.isMFARequiredError(e)) {
+    navigateToRegisterMFAScreen();
+    return;
+  }
+  // Handle any other error.
+}
+```
+
+```typescript
+// Present an UI to let the user to choose authenticator.
+// Suppose the user chose TOTP.
+
+// Present AN UI to let the user to optionally name the TOTP authenticator.
+
+const displayName = textInput.value;
+const { authenticatorID, secret } = await skygear.auth.mfa.createNewTOTP(displayName);
+
+// Present to the secret to the user and let the user
+// to add the secret to their TOTP application, such as Authy and Google Authenticator.
+
+// Present an UI to instruct the user to input the OTP.
+const otp = textInput.value;
+const { recoveryCodes } = await skygear.auth.mfa.activateTOTP(authenticatorID, otp);
+```
+
+```typescript
+// Present an UI to let the user to choose authenticator.
+// Suppose the user chose OOB.
+
+// Present an UI to instruct the user to input a phone number.
+const e164number = textInput.value;
+const { authenticatorID } = await skygear.auth.mfa.createNewOOB({ phone: e164number });
+
+// Present an UI to instruct the user to check SMS.
+
+// Present an UI to offer the user to trigger again in case the delivery failed.
+await triggerOOB(authenticatorID);
+
+// Present an UI to instruct the user to input the recevied code.
+const code = textInput.value;
+const { recoveryCodes } = await skygear.auth.mfa.activateOOB(authenticatorID, code);
+```
+
+```typescript
+// Check if recoveryCodes is present and present them to the user.
+// Ask the user to remember them and keep them safe.
+if (recoveryCodes) {
+  navigateToDisplayRecoveryCodeScreen(recoveryCodes);
+}
+```
+
+## Authentication
+
+```typescript
+try {
+  await skygear.auth.login("user@example.com", "password");
+} catch (e) {
+  // Re-raise other error.
+  if (!skygear.isMFARequiredError(e)) {
+    throw e;
+  }
+
+  let done = false;
+
+  const bearerToken = await getBearerToken();
+  if (bearerToken) {
+    try {
+      await skygear.auth.mfa.authenticateWithBearerToken(bearerToken);
```

Comment:
Do we save the authentication state (i.e. email & password) internally in SDK?

Reply:
The only things we store are

- access token
- This newly introduced MFA bearer token

Why would you ask this question?

Reply:
When I see an error occured during the `login` call, I assume the login failed and no access token is returned. So I'm thinking how the password credentials is stored in the MFA process.

Reply:
The MFA token is included in the error.

Reply:
Yes I understand, but the error is not passed to MFA function? (e.g. `authenticateWithTOTP`)

Reply:
Let me add a paragraph to explain how the client SDK should interact with the error.